protocol/smart-batching.md

```diff
@@ -0,0 +1,78 @@
+    - The Celo sequencer will run a TEE that first posts its transaction batches to a fast BFT consensus (run by Espresso Network) and only posts batches to the L1 that were already confirmed by this BFT consensus. This allows users and protocols (exchanges, bridges, etc) to obtain rollup soft confirmations (pre-L1 finality) from the fast BFT consensus that are more secure than confirmations coming from a regular centralized sequencer. Other OP stack chains are interested in adding this protective layer as well.
+- L1 directly verifies that the sequenced batches have been confirmed by an external layer (e.g. Espresso) instead of relying on a TEE.
+- Enabling various decentralized sequencing policies, by deploying an inbox contract with logic for determining which sequencer is eligible to post any given batch (similar to [RFP](https://github.com/ethereum-optimism/ecosystem-contributions/issues/63) that Espresso completed for OP Labs)
+
+## Problem Statement + Context
+
+It is desirable for many chains to customize the behavior of the OP stack. This is always possible
+by forking and modifying the derivation pipeline, but it is desirable to be able to run the canonical stack, with custom behavior as plugins. OP stack natively supports customization of certain aspects of the rollup (such as alt-DA layers) but lacks sufficient customizability to deploy new batching and sequencing policies.
+
+A simple way for rollups to deploy custom batch posting rules without needing to run a fork of
+validators and RPC nodes is to deploy a *batch inbox* contract on L1, which acts as a pre-filter to
+the derivation pipeline, rejecting any batches that don't satisfy the custom rules while
+allowing the derivation pipeline to ingest the remaining batches and execute them as normal.
+
+Unfortunately, this is not quite possible with the current derivation pipeline. While the inbox
+address can be configured to the address of a contract, the contract cannot stop batches from being ingested by the derivation pipeline, because **the derivation pipeline does not check revert status of transactions**. Thus, even if the contract reverts, the reverting transactions can still end up included in an L1 block, and the derivation pipeline will still parse and execute their calldata or
+blob data.
+
+## Proposed Solution
+
+Modify the L1 derivation stage of the pipeline (`calldata_source` and `blob_data_source`) to fetch receipts for each L1 block they process, and skip any transaction where the `status` field of the receipt is 0 (reverted transactions).
+
+### Compatibility
+
+The change is backwards compatible. If the inbox address is not a contract, transactions to it will never revert. Thus the only action necessary to opt in is to deploy a contract and upgrade to change the batch inbox address.
+
+### Resource Usage
+
+The derivation pipeline already fetches transaction receipts in several places (e.g. attributes
```

Comment:
I believe this may mean that L1 execution will have to happen in the proof system, to recreate the execution in the L1 inbox to create the receipt. We may be able to get around it by unwrapping the receipts from the receipts root in the header if we have a trusted L1 blockhash, but would want to check with @Inphi @ajsutton @clabby to understand the tradeoffs here

Reply:
> We may be able to get around it by unwrapping the receipts from the receipts root in the header if we have a trusted L1 blockhash

This was how I imagined it would work. Is there not already something like this in the proof system, since the derivation pipeline already depends on receipts (e.g. for attributes derivation)?